File: 2277-count-equal-and-divisible-pairs-in-an-array/count-equal-and-divisible-pairs-in-an-array.cpp
class Solution {
public:
    int countPairs(vector<int>& nums, int k) {
        int count=0;
        unordered_map<int,vector<int>> numIndices;
        for(int i=0;i<nums.size();++i){
            numIndices[nums[i]].push_back(i);
        }
        for (const auto[num,index]:numIndices){
            for(int i=0;i<index.size()-1;++i){
                for(int j=i+1;j<index.size();++j){
                    if((index[i]*index[j])%k==0) ++count;
                }
            }
        }
        return count;
    }
};

File: 2277-count-equal-and-divisible-pairs-in-an-array/count-equal-and-divisible-pairs-in-an-array.py
class Solution:
    def countPairs(self, nums: List[int], k: int) -> int:
        num_idx={}
        for idx,num in enumerate(nums):
            if num in num_idx:
                num_idx[num].append(idx)
            else: num_idx[num]=[idx]

        divisors = []
        for d in range(1,int(k**0.5)+1):
            if k%d==0:
                divisors.append(d)
                if d!=k/d:
                    divisors.append(k//d)
        print(divisors)
        sol = 0
        
        for idxs in num_idx.values():
            mp = Counter()
            for n in idxs:
                gcd = math.gcd(n,k)
                need = k/gcd
                sol+=mp[need]

                for d in divisors:
                    if n%d==0:
                        mp[d]+=1

        return sol